File: tests/docker-compat/js-client/tests/pubsub.js
const Redis = require('ioredis');
const { register, runTest, key, assertEqual, assertTrue, sleep } = require('./framework');

register('pubsub', async (redis, prefix) => {
  const results = [];
  const cat = 'pubsub';

  results.push(await runTest('SUBSCRIBE_PUBLISH_UNSUBSCRIBE', cat, async () => {
    const ch = key(prefix, 'ps:chan1');
    const sub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await sub.connect();

    const pub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await pub.connect();

    try {
      const messagePromise = new Promise((resolve, reject) => {
        const timeout = setTimeout(() => reject(new Error('timeout waiting for message')), 5000);
        sub.on('message', (channel, message) => {
          clearTimeout(timeout);
          resolve({ channel, message });
        });
      });

      await sub.subscribe(ch);
      await sleep(100);
      const n = await pub.publish(ch, 'hello');
      assertTrue(n >= 1, 'at least 1 subscriber should receive');

      const msg = await messagePromise;
      assertEqual('hello', msg.message);
      assertEqual(ch, msg.channel);
    } finally {
      sub.disconnect();
      pub.disconnect();
    }
  }));

  results.push(await runTest('PSUBSCRIBE_pattern', cat, async () => {
    const pattern = key(prefix, 'ps:pchan*');
    const sub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await sub.connect();

    const pub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await pub.connect();

    try {
      const messagePromise = new Promise((resolve, reject) => {
        const timeout = setTimeout(() => reject(new Error('timeout waiting for pmessage')), 5000);
        sub.on('pmessage', (pat, channel, message) => {
          clearTimeout(timeout);
          resolve({ pattern: pat, channel, message });
        });
      });

      await sub.psubscribe(pattern);
      await sleep(100);
      await pub.publish(key(prefix, 'ps:pchan_test'), 'pattern_msg');

      const msg = await messagePromise;
      assertEqual('pattern_msg', msg.message);
      assertEqual(pattern, msg.pattern);
    } finally {
      sub.disconnect();
      pub.disconnect();
    }
  }));

  results.push(await runTest('multiple_channel_subscription', cat, async () => {
    const ch1 = key(prefix, 'ps:multi1');
    const ch2 = key(prefix, 'ps:multi2');
    const sub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await sub.connect();

    const pub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await pub.connect();

    try {
      const messages = [];
      const messagePromise = new Promise((resolve, reject) => {
        const timeout = setTimeout(() => reject(new Error('timeout waiting for messages')), 5000);
        sub.on('message', (channel, message) => {
          messages.push({ channel, message });
          if (messages.length === 2) {
            clearTimeout(timeout);
            resolve(messages);
          }
        });
      });

      await sub.subscribe(ch1, ch2);
      await sleep(100);
      await pub.publish(ch1, 'msg1');
      await pub.publish(ch2, 'msg2');

      const msgs = await messagePromise;
      assertEqual('msg1', msgs[0].message);
      assertEqual('msg2', msgs[1].message);
    } finally {
      sub.disconnect();
      pub.disconnect();
    }
  }));

  results.push(await runTest('message_ordering', cat, async () => {
    const ch = key(prefix, 'ps:order');
    const sub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await sub.connect();

    const pub = new Redis({
      host: redis.options.host,
      port: redis.options.port,
      password: redis.options.password,
      lazyConnect: true,
    });
    await pub.connect();

    try {
      const messages = [];
      const messagePromise = new Promise((resolve, reject) => {
        const timeout = setTimeout(() => reject(new Error('timeout waiting for 5 messages')), 5000);
        sub.on('message', (channel, message) => {
          messages.push(message);
          if (messages.length === 5) {
            clearTimeout(timeout);
            resolve(messages);
          }
        });
      });

      await sub.subscribe(ch);
      await sleep(100);
      for (let i = 0; i < 5; i++) {
        await pub.publish(ch, `msg_${i}`);
      }

      const msgs = await messagePromise;
      for (let i = 0; i < 5; i++) {
        assertEqual(`msg_${i}`, msgs[i]);
      }
    } finally {
      sub.disconnect();
      pub.disconnect();
    }
  }));

  return results;
});

File: tests/docker-compat/python-client/tests/pubsub.py
"""Category 9: Pub/Sub tests."""

import time
import threading
import redis as redis_lib
from .framework import register, run_test, key, assert_equal, assert_true


def test_pubsub(r, prefix):
    results = []
    cat = "pubsub"

    def t_subscribe_publish():
        ch = key(prefix, "ps:chan1")
        pub = redis_lib.Redis(
            host=r.connection_pool.connection_kwargs["host"],
            port=r.connection_pool.connection_kwargs["port"],
            password=r.connection_pool.connection_kwargs.get("password"),
            db=0, decode_responses=True,
        )
        ps = r.pubsub()
        try:
            ps.subscribe(ch)
            # Consume subscription confirmation
            msg = ps.get_message(timeout=3)
            assert_true(msg is not None and msg["type"] == "subscribe", "should get subscribe confirmation")

            time.sleep(0.1)
            n = pub.publish(ch, "hello")
            assert_true(n >= 1, "at least 1 subscriber should receive")

            msg = ps.get_message(timeout=3)
            assert_true(msg is not None, "should receive message")
            assert_equal("hello", msg["data"])
            assert_equal(ch, msg["channel"])
        finally:
            ps.unsubscribe(ch)
            ps.close()
            pub.close()
    results.append(run_test("SUBSCRIBE_PUBLISH_UNSUBSCRIBE", cat, t_subscribe_publish))

    def t_psubscribe():
        pattern = key(prefix, "ps:pchan*")
        pub = redis_lib.Redis(
            host=r.connection_pool.connection_kwargs["host"],
            port=r.connection_pool.connection_kwargs["port"],
            password=r.connection_pool.connection_kwargs.get("password"),
            db=0, decode_responses=True,
        )
        ps = r.pubsub()
        try:
            ps.psubscribe(pattern)
            msg = ps.get_message(timeout=3)
            assert_true(msg is not None and msg["type"] == "psubscribe", "should get psubscribe confirmation")

            time.sleep(0.1)
            pub.publish(key(prefix, "ps:pchan_test"), "pattern_msg")

            msg = ps.get_message(timeout=3)
            assert_true(msg is not None, "should receive pattern message")
            assert_equal("pattern_msg", msg["data"])
            assert_equal(pattern, msg["pattern"])
        finally:
            ps.punsubscribe(pattern)
            ps.close()
            pub.close()
    results.append(run_test("PSUBSCRIBE_pattern", cat, t_psubscribe))

    def t_multi_channel():
        ch1 = key(prefix, "ps:multi1")
        ch2 = key(prefix, "ps:multi2")
        pub = redis_lib.Redis(
            host=r.connection_pool.connection_kwargs["host"],
            port=r.connection_pool.connection_kwargs["port"],
            password=r.connection_pool.connection_kwargs.get("password"),
            db=0, decode_responses=True,
        )
        ps = r.pubsub()
        try:
            ps.subscribe(ch1, ch2)
            # Consume both subscription confirmations
            ps.get_message(timeout=3)
            ps.get_message(timeout=3)

            time.sleep(0.1)
            pub.publish(ch1, "msg1")
            pub.publish(ch2, "msg2")

            msg1 = ps.get_message(timeout=3)
            assert_true(msg1 is not None, "should receive msg1")
            assert_equal("msg1", msg1["data"])

            msg2 = ps.get_message(timeout=3)
            assert_true(msg2 is not None, "should receive msg2")
            assert_equal("msg2", msg2["data"])
        finally:
            ps.unsubscribe(ch1, ch2)
            ps.close()
            pub.close()
    results.append(run_test("multiple_channel_subscription", cat, t_multi_channel))

    def t_ordering():
        ch = key(prefix, "ps:order")
        pub = redis_lib.Redis(
            host=r.connection_pool.connection_kwargs["host"],
            port=r.connection_pool.connection_kwargs["port"],
            password=r.connection_pool.connection_kwargs.get("password"),
            db=0, decode_responses=True,
        )
        ps = r.pubsub()
        try:
            ps.subscribe(ch)
            ps.get_message(timeout=3)

            time.sleep(0.1)
            for i in range(5):
                pub.publish(ch, f"msg_{i}")

            for i in range(5):
                msg = ps.get_message(timeout=3)
                assert_true(msg is not None, f"should receive msg_{i}")
                assert_equal(f"msg_{i}", msg["data"])
        finally:
            ps.unsubscribe(ch)
            ps.close()
            pub.close()
    results.append(run_test("message_ordering", cat, t_ordering))

    return results


register("pubsub", test_pubsub)
